### ISSUE TITLE
Proposal: Declaration Expressions

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#254]
One feature considered and ultimately not added to C# 6 was declaration expressions. A declaration expression is a new expression of the form

> _expression_:
> &emsp;&emsp;&emsp;&emsp;&emsp;_local-variable-type_ _local-variable-declarator_

Such an expression resolves to an _lvalue_ for a fresh local variable that is definitely assigned only if the declarator contains an initializer. In addition, we would support fresh variables for `out` arguments:

> _argument-value_:
> &emsp;&emsp;&emsp;&emsp;&emsp;`out` _local-variable-type_ _identifier_

We need to precisely define the scope of these variables, and perhaps require that they be initialized at the point of declaration.

**Comment by paulomorgado** [comment]
(Sorry, I don't know of any other way of being CCed when someone replies to this issue. I'll remove it later)

**Comment by paulomorgado** [comment]
(Sorry, I don't know of any other way of being CCed when someone replies to this issue. I'll remove it later)

**Comment by Miista** [comment]
There's a subscribe button at the top right of the page.

**Comment by paulomorgado** [comment]
@Miista, not if you have subscribed to the repo.

**Comment by airbreather** [comment]
Hi,

I'd just like to chime in my support for this feature and note some existing syntax that having something like this would make look much nicer / cleaner... guessing you've probably seen something like this before, but might as well:

```csharp
IEnumerable<Foo> fooSequence = new Foo[0]; // just for demo
int someThreshold = 110; // just for demo

// ...
fooSequence = fooSequence.Where(foo =>
                                {
                                    int parsedBar;
                                    return Int32.TryParse(foo.BarString, out parsedBar) &&
                                           parsedBar < someThreshold;
                                });
```
versus this, sans curly braces:
```csharp
IEnumerable<Foo> fooSequence = new Foo[0]; // just for demo
int someThreshold = 110; // just for demo

// ...
fooSequence = fooSequence.Where(foo => Int32.TryParse(foo.BarString, out int parsedBar) &&
                                       parsedBar < someThreshold);
```
I can get it to look *almost* the same without declaration expressions, but it's subtly broken in certain circumstances (well, not literally as written, but if the IEnumerable&lt;Foo&gt; were instead implicitly typed with "var" and there was a .AsParallel() or something):
```csharp
IEnumerable<Foo> fooSequence = new Foo[0]; // just for demo
int someThreshold = 110; // just for demo

// ...
int parsedBar;
fooSequence = fooSequence.Where(foo => Int32.TryParse(foo.BarString, out parsedBar) &&
                                       parsedBar < someThreshold);
```

**Comment by bondsbw** [comment]
I closed #3664 after determining it was a subset of this proposal.  It may provide an additional scenario to consider.

**Comment by alrz** [comment]
scope of variables would be the same as pattern variables, I presume.

**Comment by gafter** [comment]
See #6183 which is a subset of this but might be done sooner than this.

**Comment by alrz** [comment]
I assume that we can use declaration expressions in also ref args. :)
```cs
F(ref int x = null);
```
So there would be no assumption on the default value passed to the argument.

**Comment by NetMage** [comment]
It is really hard to tell where some of these proposals go it seems - there are no closing or summary notes on closed issues, and I assume an Open issue means (e.g. general declaration expressions) wasn't included in C# 7.0, but it seems like I should have to go to the C# design notes to tell. Perhaps it shouldn't be possible to close an issue without putting in a comment.

(BTW, is there a meta discussion area this really should be in?)

**Comment by gafter** [comment]
Declaration expressions, as added in C# 7, are discussed in https://github.com/dotnet/csharplang/issues/365 . Related to this is a proposal for sequence expressions at https://github.com/dotnet/csharplang/issues/377 , which is not currently championed in the LDM. Closing this issue.
